ebpf/throughput/main.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/cilium/ebpf/rlimit"
+	"github.com/ddosify/alaz/k8s"
+	"github.com/ddosify/alaz/log"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type ThroughputEventBpf struct {
+	Timestamp uint64
+	Size      uint32
+	SPort     uint16
+	DPort     uint16
+	SAddr     [16]byte
+	DAddr     [16]byte
+}
+
+// for user space
+type ThroughputEvent struct {
+	Timestamp uint64
+	Size      uint32
+	SPort     uint16
+	DPort     uint16
+	SAddr     string
+	DAddr     string
+}
+
+// $BPF_CLANG and $BPF_CFLAGS are set by the Makefile.
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc $BPF_CLANG -cflags $BPF_CFLAGS bpf throughput.c -- -I../headers
+
+const THROUGHPUT_EVENT = "throughput_event"
+
+var NODE_NAME = os.Getenv("NODE_NAME")
+
+func (e ThroughputEvent) Type() string {
+	return THROUGHPUT_EVENT
+}
+
+// returns when the program is loaded
+func DeployAndWait(ctx context.Context, ch chan interface{}, eventChan <-chan interface{}) {
+	// Allow the current process to lock memory for eBPF resources.
+	if err := rlimit.RemoveMemlock(); err != nil {
+		log.Logger.Fatal().Err(err).Msg("failed to remove memlock limit")
+	}
+
+	// Load pre-compiled programs and maps into the kernel.
+	objs := bpfObjects{}
+	if err := loadBpfObjects(&objs, nil); err != nil {
+		fmt.Println(err.Error())
+		log.Logger.Fatal().Err(err).Msg("loading objects")
+	}
+	defer objs.Close()
+
+	ticker := time.NewTicker(time.Millisecond)
+	defer ticker.Stop()
+
+	time.Sleep(time.Second)
+
+	// Set up network interfaces for the first time, then do it again on pod events
+	if err := setFiltersOnCiliumInterfaces(objs); err != nil {
+		log.Logger.Fatal().Err(err).Msg("failed to set up filters on cilium interfaces")
+	}
+
+	go watchNetworkInterfaces(ctx, objs, eventChan)
+
+	throughputEventReader, err := ringbuf.NewReader(objs.ThroughputEvents)
+	if err != nil {
+		log.Logger.Fatal().Err(err).Msg("failed to create ringbuf reader")
+	}
+	defer func() {
+		log.Logger.Info().Msg("closing throughputEventReader ringbuf reader")
+		throughputEventReader.Close()
+	}()
+
+	readDone := make(chan struct{})
+	go func() {
+		for {
+			read := func() {
+				record, err := throughputEventReader.Read()
+				if err != nil {
+					log.Logger.Warn().Err(err).Msg("error reading from ringbuf")
+				}
+
+				if record.RawSample == nil || len(record.RawSample) == 0 {
+					return
+				}
+
+				bpfEvent := (*ThroughputEventBpf)(unsafe.Pointer(&record.RawSample[0]))
+
+				go func() {
+					ch <- ThroughputEvent{
+						Timestamp: bpfEvent.Timestamp,
+						Size:      bpfEvent.Size,
+						SPort:     bpfEvent.SPort,
+						DPort:     bpfEvent.DPort,
+						SAddr:     fmt.Sprintf("%d.%d.%d.%d", bpfEvent.SAddr[0], bpfEvent.SAddr[1], bpfEvent.SAddr[2], bpfEvent.SAddr[3]),
+						DAddr:     fmt.Sprintf("%d.%d.%d.%d", bpfEvent.DAddr[0], bpfEvent.DAddr[1], bpfEvent.DAddr[2], bpfEvent.DAddr[3]),
+					}
+				}()
+			}
+
+			select {
+			case <-readDone:
```

Comment:
Can this just be
```suggestion
			case <-ctx.Done():
```

Reply:
Probably. I'll try it out like that. This is how their other eBPF program readers were structured, so I just followed their lead, but it does seem unnecessary to do it this way